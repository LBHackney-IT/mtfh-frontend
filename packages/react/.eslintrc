{
  "extends": "@hackney/eslint-config",
  "parserOptions": {
    "project": "./tsconfig.json"
  },
  "rules": {
    "prefer-arrow-callback": "off",
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": [
          "**/*/rollup.config.js",
          "**/*.test.ts",
          "**/*.test.tsx"
        ]
      }
    ],
    "jsx-a11y/anchor-is-valid": [
      "error",
      {
        "components": ["Link"],
        "aspects": ["invalidHref"]
      }
    ],
    "jsx-a11y/label-has-associated-control": "off"
    // "jsx-a11y/label-has-associated-control": [
    //   2,
    //   {
    //     "labelComponents": ["Label"],
    //     "labelAttributes": ["label"],
    //     "controlComponents": ["Input"],
    //     "assert": "htmlFor",
    //     "depth": 3
    //   }
    // ]
  }
}
