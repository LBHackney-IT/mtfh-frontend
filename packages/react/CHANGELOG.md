# @lbh-mtfh/react

## 1.1.0

### Minor Changes

- 37b7b3e: Initial release

### Patch Changes

- Updated dependencies [37b7b3e]
  - @lbh-mtfh/system@1.1.0
  - @lbh-mtfh/test-utils@1.1.0
