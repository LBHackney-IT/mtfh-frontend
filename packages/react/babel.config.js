const babel = require("../../babel.config")

module.exports = babel
