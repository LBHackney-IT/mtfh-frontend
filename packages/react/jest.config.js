const baseConfig = require("../../jest.config")

baseConfig.setupFilesAfterEnv.push("<rootDir>/jest.setup.ts")

module.exports = {
  ...baseConfig,
}
