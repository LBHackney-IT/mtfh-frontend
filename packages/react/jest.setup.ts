import * as LBHFrontend from "lbh-frontend"

window.LBHFrontend = LBHFrontend
