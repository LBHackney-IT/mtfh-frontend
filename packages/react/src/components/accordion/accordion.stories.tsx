import React from "react";

import { Accordion, AccordionItem } from "./accordion";

export default {
  title: "Accordion",
  component: Accordion,
};

export const Base = () => {
  return (
    <Accordion id="accordion-1">
      <AccordionItem id="item-1" title="Item 1">
        Description 1
      </AccordionItem>
      <AccordionItem id="item-2" title="Item 2">
        Description 2
      </AccordionItem>
      <AccordionItem id="item-3" title="Item 3">
        Description 3
      </AccordionItem>
    </Accordion>
  );
};

export const AccordionWithVisuallyHiddenControls = () => {
  return (
    <Accordion id="accordion-2" visuallyHideControls>
      <AccordionItem id="item-4" title="Item 1">
        Description 1
      </AccordionItem>
      <AccordionItem id="item-5" title="Item 2">
        Description 2
      </AccordionItem>
      <AccordionItem id="item-6" title="Item 3">
        Description 3
      </AccordionItem>
    </Accordion>
  );
};
