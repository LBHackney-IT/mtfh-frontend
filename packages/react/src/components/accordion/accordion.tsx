import React, { ReactElement, forwardRef, useEffect, useRef } from "react";
import mergeRefs from "react-merge-refs";

import cn from "classnames";

import { widthOverrides } from "../../utils";
import { Heading } from "../heading";
import styles from "./styles.module.scss";

import type * as Polymorphic from "@radix-ui/react-polymorphic";

export interface AccordionItemProps {
  id: string;
  title: string;
}

export type AccordionItemComponent = Polymorphic.ForwardRefComponent<
  "div",
  AccordionItemProps
>;

export const AccordionItem: AccordionItemComponent = forwardRef(function AccordionItem(
  { as: AccordionItemComp = "div", children, className, id, title },
  ref,
) {
  return (
    <AccordionItemComp ref={ref} className={cn(styles.govukAccordionSection, className)}>
      <div className={styles.govukAccordionSectionHeader}>
        <Heading as="h3" variant="h5" className={styles.govukAccordionSectionHeading}>
          <span
            className={styles.govukAccordionSectionButton}
            id={`accordion-heading-${id}`}
          >
            {title}
          </span>
        </Heading>
      </div>
      <div
        id={`accordion-content-${id}`}
        className={styles.govukAccordionSectionContent}
        aria-labelledby={`accordion-heading-${id}`}
      >
        {children}
      </div>
    </AccordionItemComp>
  );
});

type AccordionChild =
  | ReactElement<AccordionItemProps>
  | ReactElement<AccordionItemProps>[]
  | null;

export interface AccordionProps {
  id: string;
  children: AccordionChild | AccordionChild[];
  visuallyHideControls?: boolean;
  override?: number;
}

export type AccordionComponent = Polymorphic.ForwardRefComponent<"div", AccordionProps>;

export const Accordion: AccordionComponent = forwardRef(function Accordion(
  {
    as: AccordionComp = "div",
    className,
    override,
    visuallyHideControls = false,
    ...props
  },
  ref,
) {
  const localRef = useRef<HTMLElement>(null);

  useEffect(() => {
    const init = async (element: HTMLElement) => {
      const { default: LBHAccordion } = await import(
        "lbh-frontend/lbh/components/lbh-accordion/accordion"
      );
      const acc = new LBHAccordion(element);
      acc.$sections = element.querySelectorAll(`.${styles.govukAccordionSection}`);
      acc.controlsClass = styles.govukAccordionControls;
      acc.openAllClass = styles.govukAccordionOpenAll;
      acc.iconClass = styles.govukAccordionIcon;
      acc.sectionHeaderClass = styles.govukAccordionSectionHeader;
      acc.sectionHeaderFocusedClass = styles.govukAccordionSectionHeaderFocused;
      acc.sectionHeadingClass = styles.govukAccordionSectionHeading;
      acc.sectionSummaryClass = styles.govukAccordionSectionSummary;
      acc.sectionButtonClass = styles.govukAccordionSectionButton;
      acc.sectionExpandedClass = styles.govukAccordionSectionExpanded;
      acc.init();
    };
    /* istanbul ignore else */
    if (localRef.current && window !== undefined && document !== undefined) {
      init(localRef.current);
    }
  }, []);

  return (
    <AccordionComp
      className={cn(
        styles.govukAccordion,
        styles.lbhAccordion,
        { [styles.visuallyHidden]: visuallyHideControls },
        widthOverrides(override),
        className,
      )}
      data-attribute="value"
      ref={mergeRefs([localRef, ref])}
      {...props}
    />
  );
});
