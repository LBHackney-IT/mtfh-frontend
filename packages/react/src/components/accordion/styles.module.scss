@import "~lbh-frontend/lbh/base";
@import "~lbh-frontend/lbh/components/lbh-accordion/accordion";
@import "~govuk-frontend/govuk/helpers/visually-hidden";

.visually-hidden {
  .govuk-accordion__open-all {
    @include govuk-visually-hidden-focusable;
  }
}

.govuk-accordion__section[data-state="open"] {
  .govuk-accordion__section-content {
    display: block;
  }
  .govuk-accordion__icon:after {
    content: none;
  }
}

:global .js-enabled {
  :local .govuk-accordion__section-button:hover:not(.focus) {
    color: inherit;
  }
}
