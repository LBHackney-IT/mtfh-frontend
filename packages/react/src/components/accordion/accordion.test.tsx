import React from "react";
import { render, testA11y } from "@hackney/mtfh-test-utils";
import { screen, waitFor } from "@testing-library/react";

import { Accordion, AccordionItem } from "./accordion";

test("it passes a11y", async () => {
  const {
    result: { container },
  } = render(
    <Accordion id="test">
      <AccordionItem id="test-1" title="Test">
        Hello
      </AccordionItem>
      <AccordionItem id="test-2" title="Next">
        Hello
      </AccordionItem>
    </Accordion>,
  );

  await waitFor(() => expect(screen.queryAllByRole("button")).toHaveLength(3));
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it renders correctly", async () => {
  render(
    <Accordion id="test">
      <AccordionItem id="test-1" title="Test">
        Hello
      </AccordionItem>
      <AccordionItem id="test-2" title="Next">
        Hello
      </AccordionItem>
    </Accordion>,
  );

  await waitFor(() => expect(screen.queryAllByRole("button")).toHaveLength(3));

  expect(screen.getAllByRole("heading")).toHaveLength(2);
  expect(screen.getByText("Test")).toBeInTheDocument();
});
