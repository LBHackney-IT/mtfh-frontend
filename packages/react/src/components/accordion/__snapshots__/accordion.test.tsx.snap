// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it passes a11y 1`] = `
<div>
  <div
    class="govukAccordion lbhAccordion"
    data-attribute="value"
    id="test"
  >
    <div
      class="govukAccordionControls"
    >
      <button
        aria-expanded="false"
        class="govukAccordionOpenAll"
        type="button"
      >
        Open all
        <span
          class="govuk-visually-hidden"
        >
           sections
        </span>
      </button>
    </div>
    <div
      class="govukAccordionSection"
    >
      <div
        class="govukAccordionSectionHeader"
      >
        <h3
          class="lbh-heading-h5 govukAccordionSectionHeading"
        >
          <button
            aria-controls="test-content-1"
            aria-expanded="false"
            class="govukAccordionSectionButton"
            id="accordion-heading-test-1"
            type="button"
          >
            Test
            <span
              aria-hidden="true"
              class="govukAccordionIcon"
            />
          </button>
        </h3>
      </div>
      <div
        aria-labelledby="accordion-heading-test-1"
        class="govukAccordionSectionContent"
        id="accordion-content-test-1"
      >
        Hello
      </div>
    </div>
    <div
      class="govukAccordionSection"
    >
      <div
        class="govukAccordionSectionHeader"
      >
        <h3
          class="lbh-heading-h5 govukAccordionSectionHeading"
        >
          <button
            aria-controls="test-content-2"
            aria-expanded="false"
            class="govukAccordionSectionButton"
            id="accordion-heading-test-2"
            type="button"
          >
            Next
            <span
              aria-hidden="true"
              class="govukAccordionIcon"
            />
          </button>
        </h3>
      </div>
      <div
        aria-labelledby="accordion-heading-test-2"
        class="govukAccordionSectionContent"
        id="accordion-content-test-2"
      >
        Hello
      </div>
    </div>
  </div>
</div>
`;
